transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java

```diff
@@ -0,0 +1,202 @@
+    }
+
+    @Test
+    public void testWithVoidPromise() {
+        writeQueue.add(cat, voidPromise);
+        writeQueue.add(mouse, voidPromise);
+        writeQueue.add(empty, voidPromise);
+        assertQueueSize(8, false);
+        assertEquals("catm", writeQueue.remove(4, newPromise()).toString(Charset.defaultCharset()));
+        assertQueueSize(4, false);
+        assertEquals("ouse", writeQueue.remove(4, newPromise()).toString(Charset.defaultCharset()));
+        assertQueueSize(0, true);
+    }
+
+    @Test
+    public void testAggregateWithPartialRead() {
+        writeQueue.add(cat, catPromise);
+        writeQueue.add(mouse, mouseListener);
+        DefaultChannelPromise aggregatePromise = newPromise();
+        assertEquals("catm", writeQueue.remove(4, aggregatePromise).toString(Charset.defaultCharset()));
+        assertQueueSize(4, false);
+        assertFalse(catPromise.isSuccess());
+        assertFalse(mouseDone);
+        aggregatePromise.trySuccess();
+        assertTrue(catPromise.isSuccess());
+        assertFalse(mouseDone);
+
+        aggregatePromise = newPromise();
+        assertEquals("ouse", writeQueue.remove(Integer.MAX_VALUE, aggregatePromise).toString(Charset.defaultCharset()));
+        assertQueueSize(0, true);
+        assertFalse(mouseDone);
+        aggregatePromise.trySuccess();
+        assertTrue(mouseSuccess);
+    }
+
+    @Test
+    public void testReadExactAddedBufferSizeReturnsOriginal() {
+        writeQueue.add(cat, catPromise);
+        writeQueue.add(mouse, mouseListener);
+        DefaultChannelPromise aggregatePromise = newPromise();
+        assertSame(cat, writeQueue.remove(3, aggregatePromise));
+        assertFalse(catPromise.isSuccess());
+        aggregatePromise.trySuccess();
+        assertTrue(catPromise.isSuccess());
+
+        aggregatePromise = newPromise();
+        assertSame(mouse, writeQueue.remove(5, aggregatePromise));
+        assertFalse(mouseDone);
+        aggregatePromise.trySuccess();
+        assertTrue(mouseSuccess);
+    }
+
+    @Test
+    public void testReadEmptyQueueReturnsEmptyBuffer() {
+        assertQueueSize(0, true);
+        DefaultChannelPromise aggregatePromise = newPromise();
+        assertEquals(0, writeQueue.remove(0, aggregatePromise).readableBytes());
+        assertEquals(0, writeQueue.remove(0, aggregatePromise).readableBytes());
+        assertQueueSize(0, true);
+    }
+
+    @Test
+    public void testReleaseAndFailAll() {
+        writeQueue.add(cat, catPromise);
+        writeQueue.add(mouse, mouseListener);
+        RuntimeException cause = new RuntimeException("ooops");
+        writeQueue.releaseAndFailAll(cause);
+        DefaultChannelPromise aggregatePromise = newPromise();
+        assertQueueSize(0, true);
+        assertEquals(0, cat.refCnt());
+        assertEquals(0, mouse.refCnt());
+        assertSame(cause, catPromise.cause());
+        assertEquals(0, writeQueue.remove(Integer.MAX_VALUE, aggregatePromise).readableBytes());
```

Comment:
Seems like we verify the same thing on the line below (readableBytes() == 0)...consider refactoring/removing duplication.